File: src/Wording.Core/Learning/SpacedRepetitionScheduler.cs
namespace Wording.Core.Learning;

/// <summary>
/// The SM-2 (SuperMemo 2) algorithm as a pure function: it computes the next review
/// state from the previous state and a grade.
/// </summary>
public static class SpacedRepetitionScheduler
{
    /// <summary>Interval after the first successful review.</summary>
    const double FirstIntervalDays = 1.0;

    /// <summary>Interval after the second successful review.</summary>
    const double SecondIntervalDays = 6.0;

    /// <summary>
    /// After <see cref="ReviewGrade.Again"/> a word comes back shortly, not instantly -
    /// otherwise it would be permanently the most overdue word and would block the rotation.
    /// </summary>
    static readonly TimeSpan RelearnDelay = TimeSpan.FromMinutes(10);

    public static ReviewState Apply(ReviewState current, ReviewGrade grade, DateTimeOffset now)
    {
        ArgumentNullException.ThrowIfNull(current);

        var quality = (int)grade;
        var easeFactor = NextEaseFactor(current.EaseFactor, quality);

        // SM-2 treats anything below 3 as a failed recall - the repetition count restarts.
        if (quality < 3)
        {
            return current with
            {
                Repetitions = 0,
                IntervalDays = 0,
                EaseFactor = easeFactor,
                DueUtc = now + RelearnDelay,
                LastReviewedUtc = now,
                Lapses = current.Lapses + 1,
            };
        }

        var repetitions = current.Repetitions + 1;
        var interval = repetitions switch
        {
            1 => FirstIntervalDays,
            2 => SecondIntervalDays,
            _ => current.IntervalDays * easeFactor,
        };

        return current with
        {
            Repetitions = repetitions,
            IntervalDays = interval,
            EaseFactor = easeFactor,
            DueUtc = now + TimeSpan.FromDays(interval),
            LastReviewedUtc = now,
        };
    }

    /// <summary>
    /// The original SM-2 ease adjustment:
    /// EF' = EF + (0.1 - (5-q) * (0.08 + (5-q) * 0.02)), floored at 1.3.
    /// </summary>
    static double NextEaseFactor(double easeFactor, int quality)
    {
        var delta = 5 - quality;
        var updated = easeFactor + (0.1 - delta * (0.08 + delta * 0.02));

        return Math.Max(ReviewState.MinimumEaseFactor, updated);
    }
}

File: macos/Sources/WordingKit/SpacedRepetitionScheduler.swift
import Foundation

/// The SM-2 (SuperMemo 2) algorithm as a pure function.
///
/// A port of `Wording.Core.Learning.SpacedRepetitionScheduler` - both implementations
/// must produce identical numbers, because they work on the same data file.
public enum SpacedRepetitionScheduler {
    /// Interval after the first successful review.
    static let firstIntervalDays = 1.0

    /// Interval after the second successful review.
    static let secondIntervalDays = 6.0

    /// After `again` a word comes back shortly, not instantly - otherwise it would be
    /// permanently the most overdue word and would block the whole rotation.
    static let relearnDelay: TimeInterval = 10 * 60

    public static func apply(_ current: ReviewState, grade: ReviewGrade, now: Date) -> ReviewState {
        let quality = grade.rawValue
        let easeFactor = nextEaseFactor(current.easeFactor, quality: quality)

        // SM-2 treats anything below 3 as a failed recall.
        guard quality >= 3 else {
            return ReviewState(
                repetitions: 0,
                intervalDays: 0,
                easeFactor: easeFactor,
                dueUtc: now.addingTimeInterval(relearnDelay),
                lastReviewedUtc: now,
                lapses: current.lapses + 1
            )
        }

        let repetitions = current.repetitions + 1
        let interval: Double =
            switch repetitions {
            case 1: firstIntervalDays
            case 2: secondIntervalDays
            default: current.intervalDays * easeFactor
            }

        return ReviewState(
            repetitions: repetitions,
            intervalDays: interval,
            easeFactor: easeFactor,
            dueUtc: now.addingTimeInterval(interval * .day),
            lastReviewedUtc: now,
            lapses: current.lapses
        )
    }

    /// The original SM-2 formula: EF' = EF + (0.1 - (5-q) * (0.08 + (5-q) * 0.02)).
    static func nextEaseFactor(_ easeFactor: Double, quality: Int) -> Double {
        let delta = Double(5 - quality)
        let updated = easeFactor + (0.1 - delta * (0.08 + delta * 0.02))

        return max(ReviewState.minimumEaseFactor, updated)
    }
}
